labgrid/resource/mqtt.py

```diff
@@ -0,0 +1,76 @@
+import logging
+import threading
+from time import monotonic
+
+import attr
+
+from .common import ManagedResource, ResourceManager
+from ..factory import target_factory
+
+@attr.s(eq=False)
+class MQTTManager(ResourceManager):
+    _available = attr.ib(default=attr.Factory(set), validator=attr.validators.instance_of(set))
+    _avail_lock = attr.ib(default=threading.Lock())
+    _clients = attr.ib(default=attr.Factory(dict), validator=attr.validators.instance_of(dict))
+    _topics = attr.ib(default=attr.Factory(list), validator=attr.validators.instance_of(list))
+    _topic_lock = attr.ib(default=threading.Lock())
+    _last = attr.ib(default=0.0, validator=attr.validators.instance_of(float))
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        self.log = logging.getLogger('MQTTManager')
+
+    def _create_mqtt_connection(self, host):
+        import paho.mqtt.client as mqtt
+        client = mqtt.Client()
+        client.connect(host)
+        client.on_message = self._on_message
+        client.loop_start()
+        return client
+
+    def on_resource_added(self, resource):
+        host = resource.host
+        if host not in self._clients:
+            self._clients[host] = self._create_mqtt_connection(host)
+        self._clients[host].subscribe(resource.avail_topic)
+
+    def _on_message(self, client, userdata, msg):
+        payload = msg.payload.decode('utf-8')
+        topic = msg.topic
+        if payload == "Online":
+            with self._avail_lock:
+                self._available.add(topic)
+        if payload == "Offline":
+            if topic in self._available:
+                with self._avail_lock:
+                    self._available.remove(topic)
```

Comment:
Use `.discard(topic)`.

Reply:
Ah, sorry, i should have been more explicit: with `.remove()`, the initial check is no longer needed.

Reply:
I guess you meant with `.discard()` the check for the object to be in the set is not required. I have adjusted the PR accordingly.